### PR TITLE
Adding provisioner for Arch Linux

### LINKED ISSUE #1293: Arch Linux Support
Now that major distros such as Ubuntu, Debian, RHEL, CentOS, Fedora are supported, Arch Linux is the next most popular distro family. It is not as server grade as the aformentioned but it is stable enough for myself and many other people. The `docker` package is upgraded within 1-2 days of a new release and no third party repo is needed so installation should be straightforward. For the storage backend, AUFS is not supported by default but OverlayFS is.

### PR DESCRIPTION
**Pull request by hairyhenderson**
Fixes #1293 

Signed-off-by: Dave Henderson <dhenderson@gmail.com>

### CODE REVIEW

**Comment by nathanleclaire** [comment]
Oh nice.  I'll check it out when I get a chance.  Any recommendations for testing it?

**Comment by hairyhenderson** [comment]
Well I've been testing against my BeagleBone Black with the generic driver 😁

But it should work fine with the DO Arch image too...

**Comment by nathanleclaire** [comment]
> Well I've been testing against my BeagleBone Black with the generic driver 😁

Legit

**Comment by ehazlett** [comment]
If I test this on EC2 I get "OS Type not recognized".

```
ID=arch
PRETTY_NAME="Arch Linux"
```

**Comment by hairyhenderson** [comment]
> If I test this on EC2 I get "OS Type not recognized".

Ack! I'll check this out :disappointed:

**Comment by hairyhenderson** [comment]
@ehazlett - what I'm seeing on EC2 is this:

```
...
SSH cmd err, output: <nil>: NAME="Arch Linux"
ID=arch
PRETTY_NAME="Arch Linux"
ANSI_COLOR="0;36"
HOME_URL="https://www.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"


Couldn't set key , no corresponding struct field found
Error creating machine: OS type not recognized
You will want to check the provider to make sure the machine and associated resources were properly removed.
```

The `Couldn't set key , no corresponding struct field found` message is super-fishy... is this what you saw?

**Comment by hairyhenderson** [comment]
erg. that was a red herring. I think I figured it out...

**Comment by hairyhenderson** [comment]
Ok, pushed the fix. My `CompatibleWithHost` should've been testing the `ID` as well as `ID_LIKE`. I didn't test against enough platforms, clearly :)

@ehazlett @nathanleclaire PTAL

**Comment by hairyhenderson** [comment]
@ehazlett @nathanleclaire - would like to get this in before the #1729-ocalypse, anything to do before this can be merged?

File: libmachine/provision/arch.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	provisioner.SwarmOptions = swarmOptions
+	provisioner.AuthOptions = authOptions
+	provisioner.EngineOptions = engineOptions
+
+	if provisioner.EngineOptions.StorageDriver == "" {
+		provisioner.EngineOptions.StorageDriver = "overlay"
+	}
+
+	// HACK: since Arch does not come with sudo by default we install
+	log.Debug("installing sudo")
```
Comment (nathanleclaire) [inline libmachine/provision/arch.go]: Nitpick: Could you capitalize these log messages please?  Thanks

**Comment by nathanleclaire** [comment]
@hairyhenderson This is on the roadmap.

File: libmachine/provision/arch.go (hunk 1 above)
Comment (hairyhenderson) [inline libmachine/provision/arch.go]: Done

**Comment by nathanleclaire** [comment]
@hairyhenderson Can you please rebase to one commit?  Other than that it LGTM

**Comment by hairyhenderson** [comment]
@nathanleclaire - squashed!

**Comment by hairyhenderson** [comment]
@nathanleclaire - fixed for the changes from #1729 and re-squashed... Can this get merged soon?
/cc @ehazlett

**Comment by elebeaup** [comment]
I am interested in this PR but it is not yet merged and not included in the 0.5.0 RC2. What is missing from this PR in order to include it ?

**Comment by hairyhenderson** [comment]
@nathanleclaire - anything left here?
/cc @ehazlett

**Comment by dmp42** [comment]
@elebeaup @hairyhenderson sorry for this PR has not received enough attention yet.
Maintainers have been pretty busy with the plugins and libmachine work for 0.5
I understand the frustration, but this might not make it for 0.5 unfortunately.

We should still review this - ping @docker/machine-maintainers

**Comment by hairyhenderson** [comment]
@dmp42 - I'd really love if this could get in... I've been trying to get this in since before 0.4... @nathanleclaire's already given his LGTM above, so I'm not sure what else is holding it up...

**Comment by dmp42** [comment]
@hairyhenderson I see - I missed the LGTM.
Can you just rebase against master? With all the changes that landed, I want to make sure the build actually catch up everything.

Thanks!

**Comment by hairyhenderson** [comment]
@dmp42 - rebased!

**Comment by dmp42** [comment]
So, since @nathanleclaire LGTM-ed, and since the build is green, merge.

**Comment by hairyhenderson** [comment]
Thanks @dmp42! :)

**Comment by nathanleclaire** [comment]
Good times... I do wonder if soon we can pull out some of the systemd related logic into a struct which the systemd-based distro provisioners embed and that we can also use for Ubuntu >=15.04

**Pull request merged**